dev/src/index.ts

```diff
@@ -1359,36 +1354,47 @@ export class Firestore {
     const attempts = allowRetries ? MAX_REQUEST_RETRIES : 1;
     const callOptions = this.createCallOptions();
 
-    const gapicClient = this._clientPool.acquire();
-    const releaser = this._clientPool.createReleaser(gapicClient);
+    const result = new Deferred<NodeJS.ReadableStream>();
 
-    return this._retry(attempts, requestTag, () => {
-      return new Promise<NodeJS.ReadableStream>((resolve, reject) => {
-        try {
+    this._clientPool.run(gapicClient => {
+      // While we return the stream to the callee early, we don't want to
+      // release the GAPIC client until the callee has finished processing the
+      // stream.
+      const lifetime = new Deferred<void>();
+
+      this._retry(attempts, requestTag, async () => {
+        logger(
+          'Firestore.readStream',
+          requestTag,
+          'Sending request: %j',
+          request
+        );
+        const stream = gapicClient[methodName](request, callOptions);
+        const logStream = through2.obj(function(this, chunk, enc, callback) {
           logger(
             'Firestore.readStream',
             requestTag,
-            'Sending request: %j',
-            request
+            'Received response: %j',
+            chunk
           );
-          const stream = gapicClient[methodName](request, callOptions);
-          const logStream = through2.obj(function(this, chunk, enc, callback) {
-            logger(
-              'Firestore.readStream',
-              requestTag,
-              'Received response: %j',
-              chunk
-            );
-            this.push(chunk);
-            callback();
-          });
-          resolve(bun([stream, logStream]));
-        } catch (err) {
-          logger('Firestore.readStream', requestTag, 'Received error:', err);
-          reject(err);
-        }
-      }).then(stream => this._initializeStream(releaser, stream, requestTag));
+          this.push(chunk);
+          callback();
+        });
+
+        const resultStream = bun([stream, logStream]);
+        resultStream.on('close', lifetime.resolve);
```

Comment:
Are streams guaranteed to emit the `close` event? What happens in the case of an error?